storage/operation/multi_iterator.go

```diff
@@ -0,0 +1,100 @@
+package operation
+
+import (
+	"errors"
+
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/noop"
+)
+
+// multiIterator represents a logical concatenation of multiple iterators
+// in sequence.  multiIterator iterates items in the first iterator, and then
+// iterates items in the second iterator, etc.
+// Invariants:
+// - len(iterators) is > 0
+// - cur is [0, len(iterator)-1]
+type multiIterator struct {
+	iterators []storage.Iterator
+	cur       int // index of current iterator
+}
+
+var _ storage.Iterator = (*multiIterator)(nil)
+
+// NewMultiIterator returns an Iterator that is a logical concatenation of
+// multiple iterators in the provided sequence.  The returned iterator
+// iterates items in the first iterator, and then iterates items in
+// the second iterator, etc.
+func NewMultiIterator(iterators ...storage.Iterator) (storage.Iterator, error) {
+	if len(iterators) == 0 {
+		return noop.Iterator{}, nil
+	}
+	if len(iterators) == 1 {
+		return iterators[0], nil
+	}
+	return &multiIterator{iterators: iterators}, nil
+}
+
+// First seeks to the smallest key greater than or equal to the given key.
+func (mi *multiIterator) First() bool {
+	for i, iterator := range mi.iterators {
+		mi.cur = i
+
+		valid := iterator.First()
+		if valid {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Valid returns whether the iterator is positioned at a valid key-value pair.
+func (mi *multiIterator) Valid() bool {
+	return mi.iterators[mi.cur].Valid()
```

Comment:
Do we need to check `mi.cur` is within the array boundary? Or is there any guarantee mi.cur never go beyond bondary?

Reply:
> Do we need to check mi.cur is within the array boundary? Or is there any guarantee mi.cur never go beyond bondary?

`mi.cur` is expected to be within boundary.  It is [documented](https://github.com/onflow/flow-go/blob/8a5eee2f5deff8097f4169da290ae946a567c795/storage/operation/multi_iterator.go#L12-L14) in the code:

```
// multiIterator represents a logical concatenation of multiple iterators
// in sequence.  multiIterator iterates items in the first iterator, and then
// iterates items in the second iterator, etc.
// Invariants:
// - len(iterators) is > 0
// - cur is [0, len(iterator)-1]
```

I looked at the code again and don't see a way for `mi.cur` to be outside the expected boundary.